### PR TITLE
Desktop credentials pin 3853

### PR DESCRIPTION
**Pull request by charlag**

### CODE REVIEW

File: src/desktop/DesktopMain.ts [hunk 1]
```diff
@@ -90,7 +92,25 @@ type Components = {
 }
 const tfs = new TempFs(fs, electron, cryptoFns)
 const desktopUtils = new DesktopUtils(process.argv, tfs, electron)
-const desktopCrypto = new DesktopNativeCryptoFacade(fs, cryptoFns, tfs)
+const secretStorage = new KeytarSecretStorage()
+const argon2 = new WASMArgon2idFacade(async () => {
+	// this is kinda prone to deadlock if we use argon2 before
+	// conf is ready since wasm -> conf -> crypto -> wasm in an async way
+	const webAssetsPath = await conf.getConst(BuildConfigKey.webAssetsPath)
```
Comment (charlag) [inline src/desktop/DesktopMain.ts:99]: `getConst` should never need encryption. It is arguably completely independent from persistence of `DesktopConfig`, except that it gets defaults from the build config (should it just be hardcoded in the code? maybe)

File: src/desktop/DesktopNativeCryptoFacade.ts [hunk 2]
```diff
@@ -128,6 +134,8 @@ export class DesktopNativeCryptoFacade implements NativeCryptoFacade {
 		parallelism: number,
 		hashLength: number,
 	): Promise<Uint8Array> {
-		throw new Error("not implemented for this platform")
+		/** the fact that we don't want to define the cost parameters on four platforms makes it difficult to have the native
```
Comment (charlag) [inline src/desktop/DesktopNativeCryptoFacade.ts:137]: I feel like we just don't need to use `WASMArgon2idFacade` and should use `generateKeyFromPassphraseArgon2id` directly. It is the right level of abstraction in this case.

We also wouldn't need a dance with file fetch and whatnot because streaming or not doesn't make such a huge difference to us.

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 3]
```diff
@@ -69,11 +88,14 @@ export class DesktopNativeCredentialsFacade implements NativeCredentialsFacade {
 
 	private async maybeEncryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
 		if (encryptionMode === CredentialEncryptionMode.APP_PIN) {
-			const appPinKey = await this.resolveKeyFromAppPin()
+			const appPinKey = await this.resolveKeyFromAppPin(true)
+			if (appPinKey == null) {
+				throw new ProgrammingError("we should have generated a new key or thrown a CancelledErrorq")
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:93]: typo

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 4]
```diff
@@ -1,35 +1,136 @@
...
+			return await this.argon2idFacade.generateKeyFromPassphrase(newPw, newSalt)
+		} else {
+			const pw = await commonNativeFacade.promptForPassword(this.lang.get("credentialsEncryptionModeAppPin_label"))
+			const salt = base64ToUint8Array(storedAppPinSaltB64)
+			return await this.argon2idFacade.generateKeyFromPassphrase(pw, salt)
+		}
+	}
+
+	async getSupportedEncryptionModes(): Promise<Array<DesktopCredentialsMode>> {
+		return SUPPORTED_MODES.slice()
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:130]: if you would freeze them you probably don't have to slice but actually doesn't matter

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 5]
```diff
@@ -1,35 +1,136 @@
...
+		}
+	}
+
+	/**
+	 * if there is a salt stored, use it and a password prompt to derive the app pin key.
+	 * if there isn't, ask for a new password, generate a salt & store it, then derive the key.
+	 * @param generate whether to generate a new key if no salt is found
+	 * @return the derived 256 bit key or null if none is found and generate is false
+	 */
+	private async resolveKeyFromAppPin(generate: boolean): Promise<Aes256Key | null> {
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:110]: I know this mirrors our existing `KeyStoreFacade` behavior but I just want to point out that I think it's not a good idea to try to always give the key, because we run into fun bugs where the key disappears and we generate a new one and it doesn't match anymore. It should ideally be two methods and in "set pin" case we would do `getKey() ?? generateAndStoreKey()`.

The reason why it was a single method in other cases was probably 1. pre licc so each additional method is pain 2. avoiding authenticating user twice (I am not even sure if this would be the case)

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 6]
```diff
@@ -1,35 +1,136 @@
...
-	constructor(keyStoreFacade: DesktopKeyStoreFacade, crypto: DesktopNativeCryptoFacade) {
-		this._desktopKeyStoreFacade = keyStoreFacade
-		this._crypto = crypto
+	async decryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		// making extra sure that the mode is the right one since this comes over IPC
+		this.assertSupportedEncryptionMode(encryptionMode)
+		data = await this.maybeDecryptUsingAppPin(data, encryptionMode)
+		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
+		data = this.crypto.aes256DecryptKey(credentialsKey, data)
+		return data
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:47]: please do not reassign arguments. It is confusing and sometimes deprecated.

```suggestion
		// making extra sure that the mode is the right one since this comes over IPC
		this.assertSupportedEncryptionMode(encryptionMode)
		const dataWithoutPinEncryption = await this.maybeDecryptUsingAppPin(data, encryptionMode)
		const credentialsKey = await this.desktopKeyStoreFacade.getCredentialsKey()
		return  this.crypto.aes256DecryptKey(credentialsKey, dataWithoutPinEncryption)
```

this is one of the cases where proper naming of each steps actually pays off.

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 7]
```diff
@@ -1,35 +1,136 @@
...
 	}
 
-	async decryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
+	async encryptUsingKeychain(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
 		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const decryptedData = this._crypto.aes256DecryptKey(key, data)
-		return Promise.resolve(decryptedData)
+		this.assertSupportedEncryptionMode(encryptionMode)
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:52]: same things about argument and naming

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 8]
```diff
@@ -1,35 +1,136 @@
...
+		return data
 	}
 
-	async encryptUsingKeychain(data: Uint8Array, encryptionMode: CredentialEncryptionMode.DEVICE_LOCK): Promise<Uint8Array> {
-		// making extra sure that the mode is the right one since this comes over IPC
-		assert(encryptionMode === CredentialEncryptionMode.DEVICE_LOCK, "should not use unsupported encryption mode")
-		const key = await this._desktopKeyStoreFacade.getCredentialsKey()
-		const encryptedData = this._crypto.aes256EncryptKey(key, data)
-		return Promise.resolve(encryptedData)
+	private async maybeDecryptUsingAppPin(data: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:59]: I find `maybe` part of the name a bit confusing, like it might fail. I would probably only have `decryptUsingPin()` but if you keep this one how about `decryptingWithPinIfNeeded()`?

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 9]
```diff
@@ -1,35 +1,136 @@
...
+					const nativeFacade = await this.getCurrentCommonNativeFacade()
+					nativeFacade.showAlertDialog("invalidPassword_msg")
+					throw new CancelledError("app pin verification failed")
+				} else {
+					throw e
+				}
+			}
+			return data
+		} else {
+			// our mode is not app pin, so the app pin should not be set.
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:81]: The only reason we need it is because we always return a key from `resolveKeyFromAppPin` (see my comment there). If we change that one we wouldn't need this or this whole method.

**Review by charlag — COMMENTED** [review]
I was a bit surprised because after our last discussion I thought you would not do "pin + keychain". I am okay with either and I'm glad we have it now! (I didn't try it out yet, maybe someone else can do it?)

File: src/desktop/DesktopMain.ts (hunk 1 above)
Comment (ganthern) [reply src/desktop/DesktopMain.ts:99]: It's more of a theoretical possibility. Storing the build constants in the ts-code might make the build more difficult though - maybe we need to think about splitting the config handling at some point.

**Comment by ganthern** [comment]
It's better now, but there's still a case that I'm not sure how to solve (or if we should even worry about it). See last commit message.

**Comment by ganthern** [comment]
Things to change for this:
 * naming
 * the credentials encryption mode dialog cannot be canceled with ESC (was mobile-only before so not needed)
 * credentials went missing due to "invalidated keychain" when disabling the credentials while two windows were open. something does not migrate correctly

File: src/desktop/DesktopMain.ts (hunk 1 above)
Comment (charlag) [reply src/desktop/DesktopMain.ts]: I am not talking about all the constants, just about the default values for the config, my opinion is that they are not variable and have no business being in the build code.

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 10]
```diff
@@ -1,35 +1,120 @@
...
-		return Promise.resolve(encryptedData)
+	private async removeAppPinWrapper(dataWithAppPinWrapper: Uint8Array, encryptionMode: DesktopCredentialsMode): Promise<Uint8Array> {
+		// our mode is not app pin, so there is no wrapper to remove
+		if (encryptionMode !== CredentialEncryptionMode.APP_PIN) return dataWithAppPinWrapper
+		const appPinKey = await this.deriveKeyFromAppPin()
+		// if there is no salt, some other window might have turned off app pin
+		// between us entering this function and the user entering their pin.
+		// if we actually have pin-encrypted credentials and just lost the salt, we
+		// will get a decryption failure and delete the credentials further up the stack.
+		if (appPinKey == null) return dataWithAppPinWrapper
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:63]: we could also do `KeyPermanentlyInvalidatedError` here

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 11]
```diff
@@ -1,35 +1,120 @@
...
+
+		try {
+			return this.crypto.aesDecryptBytes(appPinKey, dataWithAppPinWrapper)
+		} catch (e) {
+			if (e instanceof CryptoError) {
+				const nativeFacade = await this.getCurrentCommonNativeFacade()
+				//noinspection ES6MissingAwait
+				nativeFacade.showAlertDialog("invalidPassword_msg")
+				throw new CancelledError("app pin verification failed")
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:72]: I think this is rather a `CredentialAuthenticationError`

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 12]
```diff
@@ -1,35 +1,120 @@
...
+import { CancelledError } from "../../api/common/error/CancelledError.js"
 
+/** the single source of truth for this configuration */
+const SUPPORTED_MODES = Object.freeze([CredentialEncryptionMode.DEVICE_LOCK, CredentialEncryptionMode.APP_PIN] as const)
+export type DesktopCredentialsMode = typeof SUPPORTED_MODES[number]
+
+/**
+ *
+ */
 export class DesktopNativeCredentialsFacade implements NativeCredentialsFacade {
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:21]: I wonder if we should write some tests for it now?

**Comment by charlag** [comment]
About the edge case, I think it's not super important and hard to reproduce but if we wanted to we could check for the salt before and after the prompt and if it changed just abort as cancelled.

If you wanted to you could go one step further and listen to changes of the config variable for salt and abort the prompt right away (with `Promise.race`)

**Comment by ganthern** [comment]
> About the edge case, I think it's not super important and hard to reproduce but if we wanted to we could check for the salt before and after the prompt and if it changed just abort as cancelled.
> 
> If you wanted to you could go one step further and listen to changes of the config variable for salt and abort the prompt right away (with `Promise.race`)

I completely forgot I built that into DesktopConfig

**Comment by charlag** [comment]
> > If you wanted to you could go one step further and listen to changes of the config variable for salt and abort the prompt right away (with `Promise.race`)
> 
> I completely forgot I built that into DesktopConfig

you probably don't even need to do it via config since `DesktopNativeCredentialsFacade` is not a per-window thing, you can just keep abort it on parallel requests

File: test/tests/desktop/credentials/DesktopCredentialsFacadeTest.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	} else {
+		return (await WebAssembly.instantiateStreaming(await fetch(path))).instance.exports
+	}
+}
+
+o.spec("DesktopNativeCredentialsFacade", () => {
+	const key = new Uint8Array([1, 2, 3])
+	const keyStoreFacade = makeKeyStoreFacade(key)
+
+	const getSubject = async () => {
```
Comment (charlag) [inline test/tests/desktop/credentials/DesktopCredentialsFacadeTest.ts:36]: why do we do this instead of usual `o.beforeEach()`?

File: src/desktop/credentials/DesktopNativeCredentialsFacade.ts [hunk 14]
```diff
@@ -96,20 +93,35 @@ export class DesktopNativeCredentialsFacade implements NativeCredentialsFacade {
...
+			commonNativeFacade.promptForNewPassword(this.lang.get("credentialsEncryptionModeAppPassword_label"), null),
+		)
 		const newAppPassSaltB64 = uint8ArrayToBase64(newSalt)
 		await this.conf.setVar(DesktopConfigKey.appPassSalt, newAppPassSaltB64)
 		return generateKeyFromPassphraseArgon2id(await this.argon2idFacade, newPw, newSalt)
 	}
 
+	private async tryWhileSaltNotChanged(pwPromise: Promise<string>): Promise<string> {
+		const commonNativeFacade = await this.getCurrentCommonNativeFacade()
+		return resolveChecked<string>(
```
Comment (charlag) [inline src/desktop/credentials/DesktopNativeCredentialsFacade.ts:114]: I am not sure why we need it, it is a confusing abstraction to me, it took me multiple minutes to understand what it does, even with a comment.

Why is it not

```ts
try {
 return await Promise.race(passwordPromise, cancellationPromise)
}  catch (e) {
  if (e instanceof CancelledError) {
      commonNativeFacade.showAlertDialog("retry_action")
  }

  throw e
```

Is it because you want to wait for the password to be done before showing the dialog?

Comment (charlag) [reply src/desktop/credentials/DesktopNativeCredentialsFacade.ts:114]: It would be more clear inline but I understand that you want to test it.

File: test/tests/desktop/credentials/DesktopCredentialsFacadeTest.ts (hunk 13 above)
Comment (ganthern) [reply test/tests/desktop/credentials/DesktopCredentialsFacadeTest.ts:36]: I guess because it's old magic code

Comment (charlag) [reply test/tests/desktop/credentials/DesktopCredentialsFacadeTest.ts:36]: fair
